mqttclients/python/pahoclientwrapper/paho_client_wrapper.py

```diff
@@ -0,0 +1,232 @@
+        self.mqtt_client.on_publish = self._handle_on_publish
+        self.mqtt_client.on_message = self._handle_on_message
+
+        self.connection_status = ConnectionStatus()
+        # A list of the results for each subscription in the request - either the granted qos, or -1 on failure
+        self.incoming_subacks = IncomingAckList[List[int]]()
+        # All other acks just return the packet id
+        self.incoming_unsubacks = IncomingAckList[int]()
+        self.incoming_pubacks = IncomingAckList[int]()
+        self.incoming_messages = IncomingMessageList()
+
+    @classmethod
+    def create_from_connection_settings(cls: type[TPahoClientWrapper], conn_sett:ConnectionSettings) -> TPahoClientWrapper:
+        # TODO: test envfile finding if filename is None
+        return cls(conn_sett)
+
+    @staticmethod
+    def error_string(mqtt_error: int) -> str:
+        return mqtt.error_string(mqtt_error)
+
+    def _handle_on_connect(
+            self, mqtt_client: mqtt.Client, userdata: Any, flags: Any, rc: int
+    ) -> None:
+        """
+        event handler for Paho on_connect events.  Do not call directly.
+        """
+        logger.info(
+            "_handle_on_connect called with status='{}'".format(mqtt.connack_string(rc))
+        )
+
+        # In Paho thread.  Save what we need and return.
+        if rc == mqtt.MQTT_ERR_SUCCESS:
+            # causes code waiting in `self.connection_status.wait_for_connected` to return
+            self.connection_status.connected = True
+        else:
+            # causes code waiting in `self.connection_status.wait_for_connected` to raise this exception
+            # causes code waiting in `self.connection_status.wait_for_disconnected` to return
+            self.connection_status.connection_error = Exception(mqtt.connack_string(rc))
+
+    def _handle_on_disconnect(
+            self, client: mqtt.Client, userdata: Any, rc: int
+    ) -> None:
+        """
+        event handler for Paho on_disconnect events.  Do not call directly.
+        """
+        # In Paho thread.  Save what we need and return.
+        logger.info(
+            "_handle_on_disconnect called with error='{}'".format(mqtt.error_string(rc))
+        )
+        # causes code waiting in `self.connection_status.wait_for_disconnected` to raise this exception
+        self.connection_status.connected = False
+
+    def _handle_on_subscribe(
+            self,
+            client: mqtt.Client,
+            userdata: Any,
+            mid: int,
+            granted_qos: Tuple[int, ...],  # tuple of ints
+            properties: Any = None,
+    ) -> None:
+        """
+        event handler for Paho on_subscribe events.  Do not call directly.
+        """
+        # In Paho thread.  Save what we need and return.
+        logger.info(
+            "Received SUBACK for mid {}, granted_qos {}".format(mid, granted_qos)
+        )
+        granted_qos_list = list(granted_qos)
```

Comment:
I'm guessing this is an MQTT5 thing I don't fully understand? Why are there (potentially) multiple QoS values?